misc/memory-connection-limits/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+///
+/// For these limits to take effect, this needs to be composed into the behaviour tree of your application.
+///
+/// If a connection is denied due to a limit, either a [`SwarmEvent::IncomingConnectionError`](libp2p_swarm::SwarmEvent::IncomingConnectionError)
+/// or [`SwarmEvent::OutgoingConnectionError`](libp2p_swarm::SwarmEvent::OutgoingConnectionError) will be emitted.
+/// The [`ListenError::Denied`](libp2p_swarm::ListenError::Denied) and respectively the [`DialError::Denied`](libp2p_swarm::DialError::Denied) variant
+/// contain a [`ConnectionDenied`](libp2p_swarm::ConnectionDenied) type that can be downcast to [`MemoryUsageLimitExceeded`] error if (and only if) **this**
+/// behaviour denied the connection.
+///
+/// If you employ multiple [`NetworkBehaviour`]s that manage connections, it may also be a different error.
+///
+/// # Example
+///
+/// ```rust
+/// # use libp2p_identify as identify;
+/// # use libp2p_swarm_derive::NetworkBehaviour;
+/// # use libp2p_memory_connection_limits as memory_connection_limits;
+///
+/// #[derive(NetworkBehaviour)]
+/// # #[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+/// struct MyBehaviour {
+///   identify: identify::Behaviour,
+///   limits: memory_connection_limits::Behaviour
+/// }
+/// ```
+pub struct Behaviour {
+    limits: MemoryUsageBasedConnectionLimits,
+    mem_tracker: ProcessMemoryUsageTracker,
+}
+
+impl Behaviour {
+    /// Sets the process memory usage threshold in bytes,
+    /// all pending connections will be dropped when the threshold is exeeded
+    pub fn new_with_max_bytes(bytes: usize) -> Self {
+        Self {
+            limits: MemoryUsageBasedConnectionLimits::default().with_max_bytes(bytes),
+            mem_tracker: ProcessMemoryUsageTracker::new(),
+        }
+    }
+
+    /// Sets the process memory usage threshold in the percentage of the total physical memory,
+    /// all pending connections will be dropped when the threshold is exeeded
+    pub fn new_with_max_percentage(percentage: f64) -> Self {
+        Self {
+            limits: MemoryUsageBasedConnectionLimits::default().with_max_percentage(percentage),
+            mem_tracker: ProcessMemoryUsageTracker::new(),
+        }
+    }
+
+    /// Sets memory usage refresh interval, default is 1s. Use `None` to always refresh
+    pub fn with_memory_usage_refresh_interval(mut self, interval: Option<Duration>) -> Self {
+        self.mem_tracker.refresh_interval = interval;
+        self
+    }
+
+    /// Updates the process memory usage threshold in bytes,
+    pub fn update_max_bytes(&mut self, bytes: usize) {
+        self.limits = MemoryUsageBasedConnectionLimits::default().with_max_bytes(bytes);
+    }
+
+    /// Updates the process memory usage threshold in the percentage of the total physical memory,
+    pub fn update_max_percentage(&mut self, percentage: f64) {
+        self.limits = MemoryUsageBasedConnectionLimits::default().with_max_percentage(percentage);
+    }
```

Comment:
I think these are redundant because a user could also just assign a new instance of the entire behaviour to the existing field :)

Reply:
Yes, they are mostly for writing tests, see https://github.com/libp2p/rust-libp2p/pull/4281#discussion_r1285730092